apps/fortuna/src/command/run.rs

```diff
@@ -136,41 +101,49 @@ pub async fn run(opts: &RunOptions) -> Result<()> {
     let metrics_registry = Arc::new(RwLock::new(Registry::default()));
     let rpc_metrics = Arc::new(RpcMetrics::new(metrics_registry.clone()).await);
 
-    let mut tasks = Vec::new();
+    let keeper_metrics: Arc<KeeperMetrics> =
+        Arc::new(KeeperMetrics::new(metrics_registry.clone()).await);
+    let keeper_private_key_option = config.keeper.private_key.load()?;
+    if keeper_private_key_option.is_none() {
+        tracing::info!("Not starting keeper service: no keeper private key specified. Please add one to the config if you would like to run the keeper service.")
+    }
+    let chains: Arc<RwLock<HashMap<ChainId, BlockchainState>>> = Default::default();
     for (chain_id, chain_config) in config.chains.clone() {
+        let keeper_metrics = keeper_metrics.clone();
+        let keeper_private_key_option = keeper_private_key_option.clone();
+        let chains = chains.clone();
         let secret_copy = secret.clone();
         let rpc_metrics = rpc_metrics.clone();
-        tasks.push(spawn(async move {
+        spawn(async move {
             let state = setup_chain_state(
                 &config.provider.address,
                 &secret_copy,
                 config.provider.chain_sample_interval,
                 &chain_id,
                 &chain_config,
-                rpc_metrics,
+                rpc_metrics.clone(),
             )
             .await;
-
-            (chain_id, state)
-        }));
-    }
-    let states = join_all(tasks).await;
-
-    let mut chains: HashMap<ChainId, BlockchainState> = HashMap::new();
-    for result in states {
-        let (chain_id, state) = result?;
-
-        match state {
-            Ok(state) => {
-                chains.insert(chain_id.clone(), state);
+            match state {
+                Ok(state) => {
+                    keeper_metrics.add_chain(chain_id.clone(), state.provider_address);
```

Comment:
it's probably better to run this for every chain before it's set up (?) wondering how we will report metrics in the case where the chain never gets initialized.